README.md
# webcontent_converter2

A fork of [webcontent_converter](https://pub.dev/packages/webcontent_converter)

Plugin for converting any webcontent, web uri to image bitmap or pdf file. This plugin use WebView on android, WKWebView on Ios and chromium for desktop support. This plugin was test for android, ios and desktop. 

## Support :

- &check; Android Minimum SDK Version: 21
- &check; IOS minimum target Version: 11
- &check; Deskop linux, windows, macos

| Android | IOS | Desktop|
| --- | --- | --- |
|  WebView | WkWebView | Puppeteer |


## Methods

`CONVERT TO IMAGE`

- filePathToImage(String path, double duration) : path is asset path and duration is delay time. This function will return Uint8List Image

Example: 

```
var bytes = await WebcontentConverter2.filePathToImage(path: "assets/receipt.html");
if (bytes.length > 0) _saveFile(bytes);
```

- webUriToImage(String uri, double duration) : uri is web uri(url) and duration is delay time. This function will return Uint8List Image

Example:

```
var bytes = await WebcontentConverter2.webUriToImage(
uri: "http://127.0.0.1:5500/example/assets/receipt.html");
if (bytes.length > 0) _saveFile(bytes);
```

- contentToImage(String content, double duration) : content is html or web content and duration is delay time. This function will return Uint8List Image

Example:

```
final content = Demo.getReceiptContent();
var bytes = await WebcontentConverter2.contentToImage(content: content);
if (bytes.length > 0) _saveFile(bytes);
```

`*** Purpose: The three above functions will help developer to get screenshot of html content as  Uint8List Image and push it to esc printer`

![Receipt screenshot](screenshots/receipt.jpeg?raw=true "Receipt")

`CONVERT TO PDF`

- filePathToPdf(String path, double duration, String savedPath, PdfMargins margins, PaperFormat format )

Example:

```
var dir = await getApplicationDocumentsDirectory();
var savedPath = join(dir.path, "sample.pdf");
var result = await WebcontentConverter2.filePathToPdf(
    path: "assets/invoice.html",
    savedPath: savedPath,
    format: PaperFormat.a4,
    margins: PdfMargins.px(top: 35, bottom: 35, right: 35, left: 35),
);
```

- webUriToPdf(String uri, double duration, String savedPath, PdfMargins margins, PaperFormat format )

Example:

```
var dir = await getApplicationDocumentsDirectory();
var savedPath = join(dir.path, "sample.pdf");
var result = await WebcontentConverter2.webUriToPdf(
    uri: "http://127.0.0.1:5500/example/assets/invoice.html",
    savedPath: savedPath,
    format: PaperFormat.a4,
    margins: PdfMargins.px(top: 35, bottom: 35, right: 35, left: 35),
);
```

- contentToPDF(String content, double duration, String savedPath, PdfMargins margins, PaperFormat format )

Example:

```
final content = Demo.getInvoiceContent();
var dir = await getApplicationDocumentsDirectory();
var savedPath = join(dir.path, "sample.pdf");
var result = await WebcontentConverter2.contentToPDF(
    content: content,
    savedPath: savedPath,
    format: PaperFormat.a4,
    margins: PdfMargins.px(top: 55, bottom: 55, right: 55, left: 55),
);
```

`*** Purpose: The three above functions will help developer to get pdf printed file of html content as. It will return a savedPath when saved successful otherwise null`

![Invoice screenshot](screenshots/invoice.pdf?raw=true "Invoice")

CHANGELOG.md
## 0.0.7+1
* add conditional imports for different platform implementations
## 0.0.7
* update to null-safety
* update dependencies
* format codes using `lint`
## 0.0.6+2

* fixed convert pdf on desktop (macos)

## 0.0.6+1

* fixed convert pdf on desktop (windows)

## 0.0.6

* add executablePath enable flexible for destkop deployment,

## 0.0.5

* add webview widget (allow to view webview in flutter app)

## 0.0.4

* add fromString to PaperFormat

## 0.0.3

* Hotfix IOS minimum version 11

## 0.0.2

* Hotfix IOS performance

## 0.0.1

* TODO: Describe initial release.


lib/webcontent_converter.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:dio/dio.dart';
import 'package:easy_logger/easy_logger.dart';
import 'package:flutter/foundation.dart';
import 'package:flutter/services.dart' show MethodChannel, rootBundle;
import 'package:flutter/widgets.dart';
import 'package:puppeteer/puppeteer.dart' as pp;

import 'page.dart';
import 'web_support.dart';
import 'webview_widget.dart';

export 'page.dart';
export 'webview_widget.dart';

// ignore: avoid_classes_with_only_static_members
/// [WebcontentConverter] will convert html, html file, web uri, into raw bytes image or pdf file
class WebcontentConverter {
  static const MethodChannel _channel = MethodChannel('webcontent_converter');

  static Future<String> get platformVersion async {
    final String version =
        await _channel.invokeMethod('getPlatformVersion') as String;
    return version;
  }

  /// ## `WebcontentConverter.logger`
  /// `allow to pretty text`
  /// #### Example:
  /// ```
  /// WebcontentConverter.logger('Your log text', level: LevelMessages.info);
  /// ```
  static final logger = EasyLogger(
    name: 'webcontent_converter',
    defaultLevel: LevelMessages.debug,
    enableBuildModes: [BuildMode.debug, BuildMode.profile, BuildMode.release],
    enableLevels: [
      LevelMessages.debug,
      LevelMessages.info,
      LevelMessages.error,
      LevelMessages.warning
    ],
  );

  /**
   * `IMAGE`
   * Convert html content, file, uri to image
   * Methods: [filePathToImage], [webUriToImage], [contentToImage]
   */

  /// ## `WebcontentConverter.filePathToImage`
  /// `this method read content from file vai path then call contentToImage`
  /// #### Example:
  /// ```
  ///  var bytes = await WebcontentConverter.filePathToImage(path: "assets/receipt.html");
  /// if (bytes.length > 0){
  ///   var dir = await getTemporaryDirectory();
  ///   var path = join(dir.path, "receipt.jpg");
  ///   File file = File(path);
  ///   await file.writeAsBytes(bytes);
  /// }
  /// ```
  static Future<Uint8List> filePathToImage({
    required String path,
    double duration = 2000,
    String? executablePath,
  }) async {
    Uint8List result = Uint8List.fromList([]);
    try {
      final String content = await rootBundle.loadString(path);

      result = await contentToImage(
        content: content,
        duration: duration,
        executablePath: executablePath,
      );
    } on Exception catch (e) {
      WebcontentConverter.logger.error("[method:filePathToImage]: $e");
      throw Exception("Error: $e");
    }
    return result;
  }

  /// ## `WebcontentConverter.webUriToImage`
  /// `This method read content from uri by using dio then call contentToImage`
  /// #### Example:
  /// ```
  /// var bytes = await WebcontentConverter.webUriToImage(uri: "http://127.0.0.1:5500/example/assets/receipt.html");
  /// if (bytes.length > 0){
  ///   var dir = await getTemporaryDirectory();
  ///   var path = join(dir.path, "receipt.jpg");
  ///   File file = File(path);
  ///   await file.writeAsBytes(bytes);
  /// }
  /// ```
  static Future<Uint8List> webUriToImage({
    required String uri,
    double duration = 2000,
    String? executablePath,
  }) async {
    Uint8List result = Uint8List.fromList([]);
    try {
      final Response<dynamic> response = await Dio().get(uri);
      final String content = response.data.toString();
      result = await contentToImage(
        content: content,
        duration: duration,
        executablePath: executablePath,
      );
    } on Exception catch (e) {
      WebcontentConverter.logger.error("[method:webUriToImage]: $e");
      throw Exception("Error: $e");
    }
    return result;
  }

  /// ## `WebcontentConverter.contentToImage`
  /// `This method use html content directly to convert html to List<Int> image`
  /// ### Example:
  /// ```
  /// final content = Demo.getReceiptContent();
  /// var bytes = await WebcontentConverter.contentToImage(content: content);
  /// if (bytes.length > 0){
  ///   var dir = await getTemporaryDirectory();
  ///   var path = join(dir.path, "receipt.jpg");
  ///   File file = File(path);
  ///   await file.writeAsBytes(bytes);
  /// }
  /// ```
  static Future<Uint8List> contentToImage({
    required String content,
    double duration = 2000,
    String? executablePath,
  }) async {
    final Map<String, dynamic> arguments = {
      'content': content,
      'duration': duration
    };
    Uint8List results = Uint8List.fromList([]);
    try {
      if (kIsWeb) {
        // TODO: web
        String blob = await WebSupport.toBlob();
        blob = blob.replaceAll("data:image/png;base64,", "");

        return base64.decode(blob);
      }
      if (Platform.isMacOS || Platform.isLinux || Platform.isWindows) {
        WebcontentConverter.logger.info("Desktop support");
        final pp.Browser browser = await pp.puppeteer.launch(
          executablePath: executablePath,
        );
        final pp.Page page = await browser.newPage();
        await page.setContent(content, wait: pp.Until.load);
        await page.emulateMediaType(pp.MediaType.print);
        final num offsetHeight = await page.evaluate<num>(
          'document.body.offsetHeight',
        );
        final num offsetWidth = await page.evaluate<num>(
          'document.body.offsetWidth',
        );
        results = await page.screenshot(
          format: pp.ScreenshotFormat.png,
          clip: pp.Rectangle.fromPoints(
            const pp.Point(0, 0),
            pp.Point(offsetWidth, offsetHeight),
          ),
          fullPage: false,
          omitBackground: true,
        );
        await page.close();
      } else {
        WebcontentConverter.logger.info("Mobile support");
        results = await _channel.invokeMethod('contentToImage', arguments)
            as Uint8List;
      }
    } on Exception catch (e) {
      WebcontentConverter.logger.error("[method:contentToImage]: $e");
      throw Exception("Error: $e");
    }
    return results;
  }

  /**
   * `PDF`
   * Convert html content, file, uri to pdf
   * Methods: [filePathToPdf], [webUriToPdf], [contentToPDF]
   */

  /// ## `WebcontentConverter.filePathToPdf`
  /// `This method read content from file vai path`
  /// #### Example:
  /// ```
  /// var dir = await getApplicationDocumentsDirectory();
  /// var savedPath = join(dir.path, "sample.pdf");
  /// var result = await WebcontentConverter.filePathToPdf(
  ///   path: "assets/invoice.html",
  ///   savedPath: savedPath,
  ///   format: PaperFormat.a4,
  ///   margins: PdfMargins.px(top: 35, bottom: 35, right: 35, left: 35),
  /// );
  ///```

  static Future<String?> filePathToPdf({
    required String path,
    double duration = 2000,
    required String savedPath,
    PdfMargins? margins,
    PaperFormat format = PaperFormat.a4,
    String? executablePath,
  }) async {
    String? result;
    try {
      final String content = await rootBundle.loadString(path);
      result = await contentToPDF(
        content: content,
        duration: duration,
        savedPath: savedPath,
        margins: margins,
        format: format,
        executablePath: executablePath,
      );
    } on Exception catch (e) {
      WebcontentConverter.logger.error("[method:filePathToPdf]: $e");
      throw Exception("Error: $e");
    }
    return result;
  }

  /// ## WebcontentConverter.webUriToPdf
  /// `This method read content from uri by using dio`
  /// #### Example:
  /// ```
  /// var dir = await getApplicationDocumentsDirectory();
  /// var savedPath = join(dir.path, "sample.pdf");
  /// var result = await WebcontentConverter.webUriToPdf(
  ///     uri: "http://127.0.0.1:5500/example/assets/invoice.html",
  ///     savedPath: savedPat,
  /// );
  /// ```
  static Future<String?> webUriToPdf({
    required String uri,
    double duration = 2000,
    required String savedPath,
    PdfMargins? margins,
    PaperFormat format = PaperFormat.a4,
    String? executablePath,
  }) async {
    String? result;
    try {
      final Response<dynamic> response = await Dio().get(uri);
      final String content = response.data.toString();
      result = await contentToPDF(
        content: content,
        duration: duration,
        savedPath: savedPath,
        margins: margins,
        format: format,
        executablePath: executablePath,
      );
    } on Exception catch (e) {
      WebcontentConverter.logger.error("[method:webUriToImage]: $e");
      throw Exception("Error: $e");
    }
    return result;
  }

  /// ## `WebcontentConverter.contentToPDF`
  /// `This method use html content directly to convert html to pdf then return path`
  /// #### Example:
  /// ```
  /// final content = Demo.getInvoiceContent();
  /// var dir = await getApplicationDocumentsDirectory();
  /// var savedPath = join(dir.path, "sample.pdf");
  /// var result = await WebcontentConverter.contentToPDF(
  ///     content: content,
  ///     savedPath: savedPath,
  ///     format: PaperFormat.a4,
  ///     margins: PdfMargins.px(top: 55, bottom: 55, right: 55, left: 55),
  /// );
  /// ```
  static Future<String?> contentToPDF({
    required String content,
    double duration = 2000,
    required String savedPath,
    PdfMargins? margins,
    PaperFormat format = PaperFormat.a4,
    String? executablePath,
  }) async {
    final PdfMargins _margins = margins ?? PdfMargins.zero;
    final Map<String, dynamic> arguments = {
      'content': content,
      'duration': duration,
      'savedPath': savedPath,
      'margins': _margins.toMap(),
      'format': format.toMap(),
    };
    WebcontentConverter.logger.info(arguments['savedPath'] as String);
    WebcontentConverter.logger.info(arguments['margins'] as Map<String, num>);
    WebcontentConverter.logger.info(arguments['format'] as Map<String, num>);
    String? result;
    try {
      if (Platform.isMacOS || Platform.isLinux || Platform.isWindows) {
        WebcontentConverter.logger.info("Desktop support");
        final pp.Browser browser = await pp.puppeteer.launch(
          executablePath: executablePath,
        );
        final pp.Page page = await browser.newPage();
        await page.setContent(
          content,
          wait: pp.Until.all([
            pp.Until.load,
            pp.Until.domContentLoaded,
            pp.Until.networkAlmostIdle,
            pp.Until.networkIdle,
          ]),
        );
        await page.pdf(
          format: pp.PaperFormat.inches(
            width: format.width,
            height: format.height,
          ),
          margins: pp.PdfMargins.inches(
            top: _margins.top,
            bottom: _margins.bottom,
            left: _margins.left,
            right: _margins.right,
          ),
          printBackground: true,
          output: File(savedPath).openWrite(),
        );
        await page.close();
        result = savedPath;
      } else if (Platform.isAndroid || Platform.isIOS) {
        WebcontentConverter.logger.info("Mobile support");
        result = await _channel.invokeMethod(
          'contentToPDF',
          arguments,
        ) as String;
      } else {
        // todo web
        result = null;
      }
    } on Exception catch (e) {
      WebcontentConverter.logger.error("[method:contentToPDF]: $e");
      throw Exception("Error: $e");
    }
    return result;
  }

  /// [WevView]
  static Widget webivew(String content, {double? width, double? height}) =>
      WebViewWidget(
        content,
        width: width,
        height: height,
      );
}
